keras_cv/models/__internal__/darknet_utils.py

```diff
@@ -0,0 +1,398 @@
+
+    def apply(x):
+        x = DarknetConvBlock(
+            filters,
+            kernel_size=3,
+            strides=2,
+            activation="leaky_relu",
+            name=f"{name}_conv1",
+        )(x)
+
+        for i in range(1, num_blocks + 1):
+            residual = x
+
+            x = DarknetConvBlock(
+                filters // 2,
+                kernel_size=1,
+                strides=1,
+                activation="leaky_relu",
+                name=f"{name}_conv{2*i}",
+            )(x)
+            x = DarknetConvBlock(
+                filters,
+                kernel_size=3,
+                strides=1,
+                activation="leaky_relu",
+                name=f"{name}_conv{2*i + 1}",
+            )(x)
+
+            if i == num_blocks:
+                x = layers.Add(name=f"{name}_out")([residual, x])
+            else:
+                x = layers.Add(name=f"{name}_add_{i}")([residual, x])
+
+        return x
+
+    return apply
+
+
+def SPPBottleneck(
+    filters, hidden_filters=None, kernel_sizes=(5, 9, 13), activation="silu", name=None
+):
+    """Spatial pyramid pooling layer used in YOLOv3-SPP
+
+    Args:
+        filters: Integer, the dimensionality of the output spaces (i.e. the number of
+            output filters in used the blocks).
+        hidden_filters: Integer, the dimensionality of the intermediate bottleneck space
+            (i.e. the number of output filters in the bottleneck convolution). If None,
+            it will be equal to filters. Defaults to None.
+        kernel_sizes: A list or tuple representing all the pool sizes used for the
+            pooling layers. Defaults to (5, 9, 13).
+        activation: Activation for the conv layers. Defaults to "silu".
+        name: the prefix for the layer names used in the block.
+
+    Returns:
+        a function that takes an input Tensor representing an SPPBottleneck.
+    """
+    if name is None:
+        name = f"spp{backend.get_uid('spp')}"
+
+    if hidden_filters is None:
+        hidden_filters = filters
+
+    def apply(x):
+        x = DarknetConvBlock(
+            hidden_filters,
+            kernel_size=1,
+            strides=1,
+            activation=activation,
+            name=f"{name}_conv1",
+        )(x)
+        x = [x]
+
+        for kernel_size in kernel_sizes:
+            x.append(
+                layers.MaxPooling2D(
+                    kernel_size,
+                    strides=1,
+                    padding="same",
+                    name=f"{name}_maxpool_{kernel_size}",
+                )(x[0])
+            )
+
+        x = layers.Concatenate(name=f"{name}_concat")(x)
+        x = DarknetConvBlock(
+            filters,
+            kernel_size=1,
+            strides=1,
+            activation=activation,
+            name=f"{name}_conv2",
+        )(x)
+
+        return x
+
+    return apply
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
```

Comment:
@fchollet @qlzh727 any issue with making these internal layers serializable?  I worry that this could break savedmodel creation in some way down the line if we need to change them.